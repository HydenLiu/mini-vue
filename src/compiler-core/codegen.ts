export function generate(ast) {
  // 生成render函数形式的字符串
  const context = createCodegenContext()
  const {push} = context
  push('return ')

  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')

  push(`function ${functionName}(${signature}){`)
  push(`return `)

  genNode(ast.codegenNode, context)
  push('}')

  return {
    code: context.code
  }
}

// 将code放到上下文context，文件全局可以访问
function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
  }

  return context
}

function genNode(node, context){
  const {push} = context
  push(`'${node.content}'`)
}







