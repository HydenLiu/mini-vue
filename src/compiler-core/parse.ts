import { NodeTypes } from './ast'

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  const context = createParserContext(content)
  return createRoot(parseChildren(context, []))
}

function parseChildren(context, ancestors) {
  const nodes: any = []

  while (!isEnd(context, ancestors)) {
    let node
    const s = context.source
    if (s.startsWith('{{')) {
      // 处理 {{message}}
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      // 处理element标签
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors)
      }
    }

    // 处理Text类型
    if (!node) {
      node = parseText(context)
    }

    nodes.push(node)
  }
  return nodes
}

function isEnd(context, ancestors) {
  // 当遇到结束标签的时候
  const s = context.source
  // </div>
  if (s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag
      if (startsWithEndTagOpen(s, tag)) {
        return true
      }
    }
  }

  // 当source没有值的时候
  return !s
}

// 处理Text类型
function parseText(context) {
  // 处理Text节点时，遇到'{{'停止截取
  let endIndex = context.source.length
  let endTokens = ['<', '{{']

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i])
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }

  const content = paseTextData(context, endIndex)

  return {
    type: NodeTypes.TEXT,
    content,
  }
}

// 处理element标签
function parseElement(context: any, ancestors) {
  const element: any = parseTag(context, TagType.Start)
  ancestors.push(element)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()

  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End)
  } else {
    throw new Error(`缺少结束标签：${element.tag}`)
  }
  return element
}

function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith('</') &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  )
}

function parseTag(context, type: TagType) {
  // 1. 解析tag
  const match: any = /^<\/?([a-z]*)/i.exec(context.source)
  const tag = match[1]

  // 2. 删除处理完成的代码
  advanceBy(context, match[0].length)
  advanceBy(context, 1)
  if (type === TagType.End) return

  return {
    type: NodeTypes.ELEMENT,
    tag,
  }
}

// 处理插值
function parseInterpolation(context) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  )

  advanceBy(context, openDelimiter.length)
  const rawContentLength = closeIndex - openDelimiter.length
  const rawContent = paseTextData(context, rawContentLength)
  const content = rawContent.trim()
  advanceBy(context, closeDelimiter.length)
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPERSSION,
      content,
    },
  }
}

// 删除处理完成的代码
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length)
}

// 截取文本
function paseTextData(context: any, length) {
  // 1. 获取content
  const content = context.source.slice(0, length)

  // 2. 推进 删除处理完成的代码
  advanceBy(context, length)
  return content
}

// 创建根节点
function createRoot(children) {
  return {
    children,
    type: NodeTypes.ROOT,
  }
}

// 创建上下文
function createParserContext(content: string): any {
  return {
    source: content,
  }
}
