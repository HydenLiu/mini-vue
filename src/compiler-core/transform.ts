import { NodeTypes } from './ast'
import { TO_DISPLAY_STRING } from './runtimeHelpers'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root)

  root.helpers = [...context.helpers.keys()] // render函数的参数
}

// 生成根节点的codegen
function createRootCodegen(root) {
  const child = root.children[0]
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode
  } else {
    root.codegenNode = root.children[0]
  }
}

// 创建transform上下文
function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    },
  }

  return context
}

// 遍历节点
function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms

  const exitFns: any = []
  // 执行transform
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    const onExit = transform(node, context)
    if (onExit) exitFns.push(onExit)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context)
      break
    default:
      break
  }

  let i = exitFns.length
  while(i--){
    exitFns[i]()
  }
}

// 遍历子节点
function traverseChildren(node: any, context: any) {
  const children = node.children

  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traverseNode(node, context)
    }
  }
}
