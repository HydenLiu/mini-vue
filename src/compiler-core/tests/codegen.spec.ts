import { generate } from '../codegen'
import { baseParse } from '../parse'
import { transform } from '../transform'

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
