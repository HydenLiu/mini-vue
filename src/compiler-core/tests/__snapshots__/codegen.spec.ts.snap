// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > interpolation 1`] = `"return function render(_ctx, _cache){return 'hi'}"`;

exports[`codegen > string 1`] = `"return function render(_ctx, _cache){return 'hi'}"`;
